[DTensor] default value for DTensor ops on non-participating devices

XilunWu opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #95852
* #97170

XilunWu committed:
[DTensor] default value for DTensor ops on non-participating devices

[ghstack-poisoned]

wanchaol commented on torch/distributed/_tensor/dispatch.py:
```diff
@@ -149,14 +149,47 @@ def operator_dispatch(
+        # input, the return value is a non-tensor (scalar) value.
+        # 2. else, a successful propagation. output_spec contains the return
+        # type (scalar, DTensor, List[DTensor])
+        def _default_ret_value(spec):
```
this is wrong, we should not rely on sharding propagation result, as it does not tell us if the result is a scalar value. Instead, we should analyze function schema(`op_call._schema`) return type, and maintain a map of return type to default values. (i.e. if return type is int, default value should be 0)

XilunWu replied:
> as it does not tell us if the result is a scalar value
Is the return value a scalar if its meta tensor has shape `torch.Size([])`?

wanchaol replied:
can you elaborate why this is related to meta tensor?

wanchaol replied:
Oh I see so you mean tensor metadata, yeah that could check if the output is a scalar or not.

wanchaol reviewed (changes requested):
see inline comments

wanchaol commented on torch/distributed/_tensor/dispatch.py:
```diff
@@ -149,14 +149,47 @@ def operator_dispatch(
+            if spec is None:
+                return torch.tensor(0)
+            elif isinstance(spec, DTensorSpec):
+                if spec.tensor_meta is None:
```
tensor_meta means tensor metadata, not meta tensor, these are two different concepts

wanchaol commented on torch/distributed/_tensor/dispatch.py:
```diff
@@ -149,14 +149,47 @@ def operator_dispatch(
+                elif not spec.tensor_meta.shape:
+                    return torch.tensor(0, dtype=spec.tensor_meta.dtype)
+                else:
+                    return torch.tensor([])
```
can you also assign dtype for empty tensor using tensor_meta?

wanchaol commented on torch/distributed/_tensor/dispatch.py:
```diff
@@ -149,14 +149,47 @@ def operator_dispatch(
+        # type (scalar, DTensor, List[DTensor])
+        def _default_ret_value(spec):
+            if spec is None:
+                return torch.tensor(0)
```
if `spec` is None, it means that this op is not returning tensor value, it would return other types, i.e. `int`, `bool`, I think we would need to communicate the non-tensor result to all ranks for this case, let's add the TODO you mentioned later to here, any default value here seems not making much sense, we should probably error out now.

XilunWu committed:
Update on "[DTensor] default value for DTensor ops on non-participating devices"




[ghstack-poisoned]

XilunWu committed:
Update on "[DTensor] default value for DTensor ops on non-participating devices"




[ghstack-poisoned]

XilunWu commented on test/distributed/_tensor/test_dtensor.py:
```diff
@@ -420,27 +420,68 @@ def test_dtensor_spec_local_shard_offset(self):
+        sub_mesh_assert_equal(
+            mesh.mesh,
+            torch.tensor(12.),
+            torch.tensor([]),  # TODO: this is actually wrong. should be torch.tensor(0.)
```
maybe we need extra handling for scalar tensor? The expected behavior in this case needs to reconsider.

wanchaol commented on torch/distributed/_tensor/dispatch.py:
```diff
@@ -149,14 +149,70 @@ def operator_dispatch(
+
+        def _default_ret_value(arg):
+            # TODO: Replace this dict of primitive date type
+            # default value with a communication
```
I think it should not be a TODO item, we should directly implement this feature, otherwise it's not correct. Try to call dist.allgather_object to gather the result on all ranks and return that instead.

If we don't do this, it would be wrong because once it get into a scalar, DTensor could never control it.

wanchaol commented on torch/distributed/_tensor/dispatch.py:
```diff
@@ -149,14 +149,70 @@ def operator_dispatch(
+
+            if (
+                (type_str == "Tensor")
+                or (type_str == "torch.Tensor")
```
basically, we can should distinguish between two types:

1. tensor /list[tensor] type -> can use dtensor.spec.tensor_meta to recover the dtype info
2. non tensor type -> need to call communication to ensure every rank in the world get the scalar value.

XilunWu committed:
Update on "[DTensor] default value for DTensor ops on non-participating devices"




[ghstack-poisoned]

XilunWu committed:
Update on "[DTensor] default value for DTensor ops on non-participating devices"




[ghstack-poisoned]

XilunWu committed:
Update on "[DTensor] default value for DTensor ops on non-participating devices"




[ghstack-poisoned]

XilunWu committed:
Update on "[DTensor] default value for DTensor ops on non-participating devices"




[ghstack-poisoned]

wanchaol commented on torch/distributed/_tensor/dispatch.py:
```diff
@@ -148,15 +150,111 @@ def operator_dispatch(
+            "bool",
+            "int",
+            "float",
+            "double",
```
there's no "double" in the native_functions schema, `float` is the only type.

wanchaol replied:
this seems not getting addressed.

XilunWu replied:
I haven't updated the PR to github but I have removed "double' and also the "(type == "float") or (type == "double")" if-branch because we don't have aten op that hits the dispatcher and also returns a scalar value.

wanchaol commented on torch/distributed/_tensor/dispatch.py:
```diff
@@ -148,15 +150,111 @@ def operator_dispatch(
+                else:
+                    # non-scalar tensor
+                    return torch.tensor(
+                        [],
```
the indention here seems.. weird, it should definitely fit in a single line, you might want to check if your VSCode have some issue (maybe you have some ufmt that format files to line-length 80 only

wanchaol commented on torch/distributed/_tensor/dispatch.py:
```diff
@@ -148,15 +150,111 @@ def operator_dispatch(
+        # tensors with correct dtype.
+
+        def _default_tensor(spec: Optional[DTensorSpec]):
+            assert spec is not None, (
```
if you are asserting spec is not None, we should just pass in non optional type?

XilunWu replied:
if `output_sharding.output_spec` is a `Sequence`, the `spec` variable in list comprehension `[... for spec in output_sharding.output_spec` will have type `Optional[DTensorSpec]`. Does it look better if I put type judgement in list comprehension:
```
return [
    _default_tensor(spec) if spec is not None else None  # output can't be a mix of scalar and Tensor values
    for spec in output_sharding.output_spec
]
```

wanchaol commented on torch/distributed/_tensor/dispatch.py:
```diff
@@ -148,15 +150,111 @@ def operator_dispatch(
+
+        def _default_ret_value(arg):
+            type_str = str(arg.type)
+            if type_str == "()":  # used in aten function schema for None return value
```
hmmm can you point me to an op with "()" return value type? I couldn't find it in native_functions.yaml

XilunWu replied:
my intention was, for ops like `_backward()` which returns None.
https://github.com/pytorch/pytorch/blob/cc699c56dc14af5b652a49e3044c0fe7c8ca70ee/aten/src/ATen/native/native_functions.yaml#L40-L43

wanchaol replied:
Oh, those ops won't hit to the dispatcher, so we can safely ignore that

wanchaol commented on torch/distributed/_tensor/dispatch.py:
```diff
@@ -148,15 +150,111 @@ def operator_dispatch(
+                obj_list = [None for _ in range(dist.get_world_size())]
+                dist.all_gather_object(obj_list, None)
+                obj_list = list(filter(lambda x: x is not None, obj_list))
+                # TODO: this nested if-else statement is due to MYPY's
```
this is actually simple, just use:
```
import operator
functools.reduce(operator.and_, obj_list, True)
```

wanchaol commented on torch/distributed/_tensor/dispatch.py:
```diff
@@ -148,15 +150,111 @@ def operator_dispatch(
+                # smarter way to do it.
+                if type_str == "bool":
+                    return functools.reduce(lambda x, y: x and y, obj_list, True)  # type: ignore[arg-type]
+                elif (
```
I think a simple "else" branch would be sufficient here, as you are already in `_arg_type_is_primitive==True` condition

wanchaol commented on torch/distributed/_tensor/dispatch.py:
```diff
@@ -148,15 +150,111 @@ def operator_dispatch(
+        # 1. if the return type is scalar, all gather the local result
+        # from participating devices, and reduce on the list of results.
+        # For bool type, we by default use AND to reduce;
+        # For numeric type, we by default use ADD to reduce;
```
I don't know if doing `add` for numeric type is correct or not. Could you give some example operator that behave like `add`? Otherwise I would say:
- if it's bool -> and
- else: raise RuntimeError

XilunWu replied:
You're right. Actually after checking `native_functions.yaml` I don't see any operator that returns numeric value and also dispatches to tensor computation.

wanchaol replied:
so we should delete it right? but it's still there

XilunWu replied:
Yes we should delete it. I haven't pushed the update to github yet.

wanchaol commented on torch/distributed/_tensor/dispatch.py:
```diff
@@ -148,15 +150,111 @@ def operator_dispatch(
+                    f"Sequence[DTensorSpec], got {output_sharding.output_spec}"
+                )
+                return [
+                    _default_tensor(spec)
```
actually could we do this List logic in `default_tensor` function directly? maybe rename it to sth like `default_tensor_types`, and handle the list variant inside.

XilunWu replied:
The reason I extract `default_tensor` instead of handling Tensor and List[Tensor] in one function is, I want to avoid that this logic appears twice. Otherwise:
```
default_tensor_types(output_spec):
    if (is_tensor()):
        extract dtype
        return torch.tensor(...)
    elif (is_tensor_list()):
        result = []
        for spec in output_spec:
            extract dtype
            result.append(torch.tensor(...))
        return result
```

But we can split `default_ret_value` into `default_primitive_types` and `default_tensor_types`. And in `default_tensor_types` we check it's a Tensor or List[Tensor]. For List[Tensor], we do the list comprehension and call `default_tensor`; for Tensor, call `default_tensor`.  What do you think?

wanchaol replied:
I think the current way of structing the code is too messy, we should reduce the number of util functions to make it more modular, and unify some check logics if possible.

wanchaol commented on torch/distributed/_tensor/dispatch.py:
```diff
@@ -148,15 +150,111 @@ def operator_dispatch(
+            local_results = _default_ret_value(ret_list[0])
+        elif len(ret_list) > 1:
+            # return: List[torch.Tensor]
+            local_results = [_default_ret_value(arg) for arg in ret_list]
```
I don't quite understand, if you already return a default tensor list in `_default_ret_value`, wouldn't this additional list comprehension become wrong?

wanchaol commented on torch/distributed/_tensor/dispatch.py:
```diff
@@ -174,6 +272,19 @@ def operator_dispatch(
         local_results = op_call(*local_tensor_args, **local_tensor_kwargs)
+        if (
+            (mesh is not None)
+            and (mesh.mesh.numel() < dist.get_world_size())
```
hmmm I think this condition should be the same as the submesh condition? `mesh is not None and mesh.get_coordinate() is None`.

XilunWu replied:
`mesh.get_coordinate() is None` means that `self.rank` is not a part of submesh but this if-block is for participating ranks. The if-condition means "op is operating on a submesh (not a full mesh) and the current rank is a part of submesh". For a full mesh, we don't need to consider default value, right?

wanchaol replied:
for participating rank, shouldn't we check with `mesh.get_coordinate() is not None`?

XilunWu replied:
In this else-branch, `mesh.get_coordinate() is not None` has been checked:
```
if mesh is not None and mesh.get_coordinate() is None:
    ....
else:
    ....
    if (mesh is not None) and (mesh is a submesh):
        all_gather_object()
```

wanchaol replied:
Oh we want to check if the op is running in a "submesh" circumstance here, and only communicate when it returns a scalar value, got it, this make sense then.

wanchaol commented on torch/distributed/_tensor/dispatch.py:
```diff
@@ -174,6 +272,19 @@ def operator_dispatch(
+            ret_list = op_schema.func_schema.returns
+            if (
+                (len(ret_list) == 1)
+                and (_arg_type_is_primitive(str(ret_list[0].type)))
```
I think `ret_list` is always length == 1, as function schema always have a single return (if it;s multiple return, it would return Tuple or List), are you saying for `List[Tensor]` case, it would actually have multiple return types?

XilunWu replied:
My confusion was that, in `native_functions.yaml` there're many functions that have multiple return values (e.g. search for `(Tensor, Tensor)`) but they seem to not hit the dispatcher at all. For example, aten functions like `native_dropout` return a tuple of tensors but it doesn't hit the dispatcher because it's `aten::native`. Is that correct?

wanchaol replied:
No, `native_dropout` did hit the dispatcher, if there's a `dispatch` key in the schema, it would hit the dispatcher

XilunWu replied:
I see. I wanted to see if for native_dropout (https://github.com/pytorch/pytorch/blob/cc699c56dc14af5b652a49e3044c0fe7c8ca70ee/aten/src/ATen/native/native_functions.yaml#L241-L248) whose return type is "(Tensor, Tensor)" the `ret_list` is still of length 1. So I tried:
```
dropout = torch.nn.Dropout(p=0.5)
dtensor = dropout(dtensor)
```
but the operator hitting dispatcher is actually `dropout` (https://github.com/pytorch/pytorch/blob/cc699c56dc14af5b652a49e3044c0fe7c8ca70ee/aten/src/ATen/native/native_functions.yaml#L269-L270). 

Do you know how to test with `native_dropout` function? Or, I misunderstood here. `native_dropout` is an ATen op which is called in the torch op `dropout`. Only `dropout` hits DTensor's dispatcher and `native_dropout` hits PyTorch dispatcher, is that right?

wanchaol replied:
there's no difference between DTensor dispatcher and pytorch's one. `native_dropout` is an aten op and we are only targetting on aten op, I think we registered `native_dropout` https://github.com/pytorch/pytorch/blob/60e503d4685f7cb4e7b7af5b11548c1d47d299e0/torch/distributed/_tensor/ops/pointwise_ops.py#L403 So don't consider torch op

If you don't see it, it's likely you are running on CPU instead of GPU, I remember somewhat native_dropout only get hit on GPU.

wanchaol reviewed (commented):
See comments inlined. I think you might need to configure our editor and lint the files with lintrunner (looks like the current linting you have is assuming len = 80).

wanchaol reviewed (commented):
did you actually update the PR? It looks like it's still the old one

XilunWu committed:
Update on "[DTensor] default value for DTensor ops on non-participating devices"




[ghstack-poisoned]

wanchaol commented on torch/distributed/_tensor/dispatch.py:
```diff
@@ -149,14 +151,72 @@ def operator_dispatch(
+                local_results = []
+                for s in spec:
+                    if s is not None:
+                        local_results.append(default_tensor(s))
```
nit: `local_results = [default_tensor(s) if s is not None else None for s in spec]`

XilunWu replied:
My confusion right now is, can we safely assume the return List has only Tensor objects? E.g. lstm_mps_backward (
https://github.com/pytorch/pytorch/blob/cc699c56dc14af5b652a49e3044c0fe7c8ca70ee/aten/src/ATen/native/native_functions.yaml#L7222-L7225). I checked our op registration and all multiple returns ops only return multiple Tensors.

wanchaol replied:
we don't need to worry about one corner case as we can always deal with that later when we actually add that op

XilunWu replied:
agree. then I'll just add
```
if None in local_results:
    raise NotImplementedError(...)
```

wanchaol commented on torch/distributed/_tensor/dispatch.py:
```diff
@@ -149,14 +151,72 @@ def operator_dispatch(
+                        local_results.append(default_tensor(s))
+                    else:
+                        raise NotImplementedError(
+                            f"returning a list of mixed scalar and Tensor"
```
we should just say return type {type} in DTensor op is not supported

wanchaol commented on torch/distributed/_tensor/dispatch.py:
```diff
@@ -149,14 +151,72 @@ def operator_dispatch(
+                local_results = functools.reduce(operator.and_, obj_list, True)
+            else:
+                raise NotImplementedError(
+                    f"default value for type {ret_type} on non-participating"
```
nit: `return type {type} in DTensor op is not supported`

wanchaol reviewed (commented):
almost ready, have a few last comments.

XilunWu committed:
Update on "[DTensor] default value for DTensor ops on non-participating devices"




[ghstack-poisoned]

XilunWu committed:
Update on "[DTensor] default value for DTensor ops on non-participating devices"




[ghstack-poisoned]

XilunWu committed:
Update on "[DTensor] default value for DTensor ops on non-participating devices"




[ghstack-poisoned]

XilunWu commented:
lint fixed. ready for review

XilunWu committed:
Update on "[DTensor] default value for DTensor ops on non-participating devices"




[ghstack-poisoned]

wanchaol reviewed (approved):
lgtm, thanks for addressing comments!

XilunWu commented:
@XilunWu has imported this pull request.  If you are a Meta employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D44321774).

XilunWu commented:
@pytorchmergebot merge -g

XilunWu commented:
@pytorchbot merge -g

XilunWu commented:
@pytorchmergebot merge -r master

XilunWu committed:
Update on "[DTensor] default value for DTensor ops on non-participating devices"


Differential Revision: [D44321774](https://our.internmc.facebook.com/intern/diff/D44321774)

[ghstack-poisoned]

XilunWu commented:
@XilunWu has imported this pull request.  If you are a Meta employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D44321774).

XilunWu commented:
@pytorchmergebot merge

XilunWu commented:
@pytorchmergebot merge
